app/presenters/compute_resource_presenter.rb

```diff
@@ -0,0 +1,93 @@
+class ComputeResourcePresenter
+  def self.for_cr(compute_resource, view)
+    klass = "#{compute_resource.provider.camelcase}Presenter".safe_constantize
+    Rails.logger.warn("Do not know how to present #{compute_resource.provider}") unless klass
+    klass ||= self
+    klass.new(compute_resource, view)
+  end
+
+  attr_accessor :model, :view
+  alias_method :compute_resource, :model
+
+  def initialize(compute_resource, view)
+    @model, @view = compute_resource, view
+  end
+
+  # TODO: preparation of the cache should be private in the future
+  def prepare_host_cache!(vms)
```

Comment:
Should we use a single presenter for multiple `vms` lists? If we do, we should not cache partial hosts list.
Personally I prefer a presenter per list, then `@vms` goes to initializer and the whole method can become a memoized accessor:

``` ruby
def hosts
  @hosts ||= Host.for_vm(compute_resource, vms).group_by(:uuid)
end
```